json-core/src/main/java/io/micronaut/json/JsonCodec.java

```diff
@@ -0,0 +1,94 @@
+/*
+ * Copyright 2017-2021 original authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.micronaut.json;
+
+import io.micronaut.core.annotation.AnnotationMetadata;
+import io.micronaut.core.annotation.Internal;
+import io.micronaut.core.convert.ConversionError;
+import io.micronaut.core.convert.exceptions.ConversionErrorException;
+import io.micronaut.core.type.Argument;
+import io.micronaut.json.tree.JsonNode;
+import org.reactivestreams.Processor;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.OutputStream;
+import java.nio.charset.StandardCharsets;
+import java.util.Optional;
+import java.util.function.Consumer;
+
+public interface JsonCodec {
+    <T> T readValueFromTree(JsonNode tree, Argument<T> type) throws IOException;
+
+    default <T> T readValueFromTree(JsonNode tree, Class<T> type) throws IOException {
+        return readValueFromTree(tree, Argument.of(type));
+    }
+
+    JsonNode writeValueToTree(Object value) throws IOException;
+
+    @Internal
+    default void updateValueFromTree(Object value, JsonNode tree) throws IOException {
```

Comment:
Perhaps create a separate internal interface that that this interface extends from rather than having all the methods in the public API. Or even better another interface that we cast down to in the cases where these methods are needed. Mixing internal and public methods in the same API should be avoided if possible

Reply:
going the cast route, that would require all implementors to implement the second interface. is there a way to mark the JsonCodec so that extension by users is unsupported?

Reply:
There are a few different methods here. `updateValueFromTree` is supported by jackson, but I want to avoid implementing this with the generated serializers, so the method may be removed at some point if I can replace all the uses. I've marked it experimental, instead.

I've removed `@Internal` from some of the other APIs.

I will move `newConversionError` to its own interface. It has only a single use anyway.